### ISSUE TITLE
timer_api intermittent failures on Nordic nRF

### ISSUE THREAD

**Issue by pabigot** [zephyrproject-rtos/zephyr#23106]
The same failure identified in #21813 and #22623 can occur on Nordic hardware:
```
starting test - test_timer_duration_period

Assertion failed at ZEPHYR_BASE/tests/kernel/timer/timer_api/src/main.c:64: duration_expire: interval >= 100 is false
```

Analysis suggests this is a false failure: the test measures durations in milliseconds, while the Nordic clock tracks 32 KiHz (32768 Hz) ticks that don't align to millisecond boundaries.  Since information is lost every time a 32 KiHz count is truncated to a 1000 Hz count it's possible that the difference in time measured in 32 KiHz counts is greater than 100 ms while the difference in 1000 Hz counts is less than 100 ms.  To avoid the information loss the timestamps need to be measured in hardware ticks with millisecond durations calculated directly from the difference in hardware ticks.

**Comment by pabigot** [comment]
@ioannisg @carlescufi

**Comment by pabigot** [comment]
These failures is usually particularly repeatable on the board formerly known as `nrf51_pca10028`. which also has:
```

starting test - test_timer_remaining

    Assertion failed at ../src/main.c:547: test_timer_remaining: rem_ticks <= dur_ticks / 2 is false

FAIL - test_timer_remaining
===================================================================
starting test - test_timeout_abs

    Assertion failed at ../src/main.c:570: test_timeout_abs: t2.ticks == t.ticks is false
```

**Comment by pabigot** [comment]
This is mostly fixed by #24080.  There's one additional test case where relaxation is required using `INEXACT_MS_CONVERT`.  I'm also checking the clock error on the board I can reproduce this with.

**Comment by pabigot** [comment]
The board I can reproduce this on has a 16000 ppm error in the 16M RCOSC, which is 1.5% and within the +/-5% allowed error, but is too large for the currently coded configurations.  It doesn't seem worth relaxing the test tolerance again for this.
